staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/unstructured_conversion_test.go

```diff
@@ -0,0 +1,465 @@
+		objectToConvert               runtime.Object
+		expectedErr                   error
+		expectedConvertedUnstructured *unstructured.Unstructured
+	}{
+		{
+			name:                          "convert nil object to unstructured should fail",
+			objectToConvert:               nil,
+			expectedErr:                   fmt.Errorf("unable to convert object type <nil> to Unstructured, must be a runtime.Object"),
+			expectedConvertedUnstructured: &unstructured.Unstructured{},
+		},
+		{
+			name:            "convert empty versioned object to unstructured should succeed",
+			objectToConvert: &testapigroupv1.Carp{},
+			expectedConvertedUnstructured: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Carp",
+					"metadata": map[string]interface{}{
+						"creationTimestamp": nil,
+					},
+					"spec":   map[string]interface{}{},
+					"status": map[string]interface{}{},
+				},
+			},
+		},
+		{
+			name: "convert valid versioned object to unstructured should succeed",
+			objectToConvert: &testapigroupv1.Carp{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "noxu",
+				},
+				Spec: testapigroupv1.CarpSpec{
+					Hostname: "example.com",
+				},
+			},
+			expectedConvertedUnstructured: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Carp",
+					"metadata": map[string]interface{}{
+						"creationTimestamp": nil,
+						"name":              "noxu",
+					},
+					"spec": map[string]interface{}{
+						"hostname": "example.com",
+					},
+					"status": map[string]interface{}{},
+				},
+			},
+		},
+		{
+			name:                          "convert internal object to unstructured should fail",
+			objectToConvert:               &testapigroup.Carp{},
+			expectedErr:                   fmt.Errorf("unable to convert the internal object type *testapigroup.Carp to Unstructured without providing a preferred version to convert to"),
+			expectedConvertedUnstructured: &unstructured.Unstructured{},
+		},
+	}
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			outUnstructured := &unstructured.Unstructured{}
+			err := scheme.Convert(testCase.objectToConvert, outUnstructured, nil)
+			assert.Equal(t, testCase.expectedErr, err)
+			assert.Equal(t, testCase.expectedConvertedUnstructured, outUnstructured)
+		})
+	}
+}
+
+func TestUnstructuredToObjectConversion(t *testing.T) {
+	scheme, _ := test.TestScheme()
+	testCases := []struct {
+		name                    string
+		unstructuredToConvert   *unstructured.Unstructured
+		convertingObject        runtime.Object
+		expectPanic             bool
+		expectedErrFunc         func(err error) bool
+		expectedConvertedObject runtime.Object
+	}{
+		{
+			name:                  "convert nil unstructured to versioned object will panic",
+			unstructuredToConvert: nil,
+			convertingObject:      &testapigroupv1.Carp{},
+
+			expectPanic:             true,
```

Comment:
might be worthwhile to add checks to make this an error, not a panic (in a followup, keeping this a test-only PR is fine)